### DISCUSSION TITLE
Can't get reactivity in my Cart.js Minicart

### DISCUSSION THREAD

**Discussion by james0r** [alpinejs/alpine#1181] [1. Help] [answered]
I'm passing CartJS from the window to my component (the document body), and then in my minicart i'm calling a CartJS function on CartJS that is now in my component scope.

The request is made but i can't manage to get my minicart to update after the item is removed.

Here is my minicart code:

```
<div class="minicart">
  <div class="minicart--item-container">
    You have
    <span
      class="minicart--item-count"
      style="font-weight: 600"
      x-html="cjs.item_count"
    ></span>
    in your cart!
  </div>
  <hr />
  <ul>
    <template
      x-for="(item, index, collection) in cjs.items"
      :key="index"
    >
      <li class="minicart--item">
        <img
          class="placeholder"
          height="80"
          width="80"
          x-bind:src="item.image"
          x-bind:alt="item.title"
        />
        <h1 class="title" x-text="item.title"></h1>
        <p class="price" x-text="money_without_trailing_zeros(item.price)" style="margin: 0;"></p>
        <small x-text="`Quantity:` + item.quantity" style="margin-bottom: 6px; display: inline-block;"></small>
        <p class="remove">
          <a x-on:click="CartJS.updateItem((index + 1), 0)">
            <i class="fa fa-trash-o"></i>
            Remove from cart
          </a>
        </p>
      </li>
    </template>
  </ul>
  <hr />
  <div class="minicart--subtotal">
    <p class="minicart--subtotal-title">Subtotal</p>
    <p class="minicart--subtotal-amount">$270.00 USD</p>
  </div>
  <input type="button" value="View Cart Details" />
</div>

<script>
  function money_without_trailing_zeros(money) {
    return "$" + money / 100;
  }
</script>
```

and the body element where I init my "component":

```
<!doctype html>
<html lang="{{ request.locale.iso_code }}">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <link rel="preconnect" href="https://cdn.shopify.com" crossorigin>

    <script src="https://code.jquery.com/jquery-3.6.0.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>

    <script>
      function initCartJS() {
        CartJS.init({{ cart | json }});
      }

      jQuery(function() {
        // jQuery('#{{ page_title | handle }}').on("dirtychange", function (e) {
        //   console.log( e );
        // })
        jQuery(document).on('cart.requestComplete', function(e, cart) {
          console.log( 'Im back!' );

          // const requestCompleteEvent = new CustomEvent('dirtychange', {
          //   detail: {
          //     updatedCartJS: cart
          //   }
          // });
  
          // document.querySelector('#{{ page_title | handle }}').dispatchEvent(requestCompleteEvent);

          document.querySelector('#{{ page_title | handle }}').__x.$data.cjs = cart;

        });
      })

    </script>

    <link rel="preload" href="{{ 'bundle.css' | asset_url }}" as="style">
    <link rel="preload" href="{{ 'bundle.js' | asset_url }}" as="script">

    <title>{{ page_title | escape }}</title>

    {% if page_description %}
      <meta name="description" content="{{ page_description | escape }}">
    {% endif %}

    {% comment %}
      css bundle is not loaded during the development task, instead it displays a 404 error in the web console
    {% endcomment %}
    {{ 'bundle.css' | asset_url | stylesheet_tag }}

    <!-- header hook for Shopify plugins -->
    {{ content_for_header }}
  </head>

  <body id="{{ page_title | handle }}" class="template-{{ request.page_type | handle }}" x-data="{ cjs: CartJS.cart }" x-init='initCartJS()' x-on:dirtychange="something = $event.detail.updatedCartJS">
    <div {% unless request.page_type == 'captcha' %}id="app"{% endunless %}>
      {% section 'header' %}

      {% render 'mini-cart' %}

      <main id="main" role="main">
        {{ content_for_layout }}
      </main>

      {% section 'footer' %}
    </div>

    <!-- webpack bundle -->
    {{ 'bundle.js' | asset_url | script_tag }}

  </body>
</html>
```

Update:

I've now tried method 2 and 3 from this answer

[https://stackoverflow.com/questions/63233506/refreshing-data-after-the-fact-in-alpinejs](https://stackoverflow.com/questions/63233506/refreshing-data-after-the-fact-in-alpinejs)

But neither updates my view or updates the info available in the chrome alpine devtools.

inspecting the element with

`document.querySelector('body').__x.$data.cjs`

for method 2 does show the updated values though.

**Comment by KevinBatdorf** [comment]
I don't know Cart.js but shouldnt this be: `x-init='cartjs.init()'` with lowercase `cartjs`?

if not, can you set up demo on CodePen?

**Reply by james0r** [reply]
The CartJS docs call for init'ing with:

```
        <!-- Initialise Cart.js once the page has loaded -->
        <script type="text/javascript">
            jQuery(function() {
                CartJS.init({{ cart | json }});
            });
        </script>
```

I think this CodePen decently reproduces my problem.

https://codepen.io/James0r/pen/BaQENaa

[CartJS](https://cartjs.org/) lives on the window. I don't believe theres a way i can--or should--bring my CartJS into my component, so from what I understand, I have to use CustomEvents to trigger a refresh of my component once I update CartJS and get my response back.

**Reply by KevinBatdorf** [reply]
That codepen doesn't seem to have cartJS on it?

> I have to use CustomEvents to trigger a refresh

Sounds right, but you can usually hook into those events from with your component and update your component state accordingly

**Reply by james0r** [reply]
Well I cant use CartJS on codepen cause it's Shopify specific. I'll update my question with the method 3 approach I took from the SE answer.

Maybe someone using Alpine on Shopify has worked this out already.

**Reply by SimoTod** [reply]
You can use cartJS outside shopify, as far as the documentation says: https://cartjs.org/pages/guide

I'm not at my laptop right now but i can send you an example tomorrow.

**Reply by james0r** [reply]
> You can use cartJS outside shopify, as far as the documentation says: https://cartjs.org/pages/guide

I'm not seeing where it says that. Seems like the whole library is an abstraction for the Shopify cart API.

**Reply by SimoTod** [reply]
Well, you won't get the updated items back but nobody stops you from instantiating it in a codepen.
See https://codepen.io/SimoTod/pen/WNoWLNY

The cart is instantiated at the top using jquery (I'm not familiar with the library so I just followed the doc).

The first component has the cart object and a control variable storing the timestamp. On init, it sets a listener for the cart event to update the control variable (which will trigger the refresh so you should read updated params from the cart object), if you don't get the updated cart straightaway, add another listener for the cart.ready event and update the variable again. (Note, alpine has a better syntax to create listeners, @event.window, but it doesn't support event names with dots so I used x-init here).

The second alpine component just uses the CartJS API to add an item (this call fails since there is no backend but you still get the events).

Using that logic, you should be able to get alpine components to react on cart changes.

**Comment by james0r** [comment, marked as answer]
I ended up scrapping CartJS and just using the shopify cart API and fetch. If anyone is interested, this ended up being a lot easier to manage.

I took some inspiration from this simple fetch + Alpine demo.

https://codepen.io/hugodf/pen/OJVwbbZ
